data_model/src/predicate.rs

```diff
@@ -204,13 +204,181 @@ impl Default for PredicateBox {
+///
+/// # Examples
+///
+/// ```rust
+/// # use iroha_data_model::test_predicate_symbol_conformity;
+/// fn test_conformity() {
+///     test_predicate_symbol_conformity!(bool; true, false);
+/// }
+/// ```
+///
+/// With `shallow_eq`:
+///
+/// ```rust
+/// # use iroha_data_model::{PredicateSymbol, test_predicate_symbol_conformity};
+/// # use std::ops::ControlFlow;
+///
+/// #[derive(Clone, PartialEq)]
+/// enum Check {
+///    Good,
+///    // Encapsulates reason for badness which
+///    // doesn't behave associatively
+///    // (but if we ignore it, Check as a whole does)
+///    Bad(String),
+/// }
+///
+/// impl core::ops::Not for Check {
+///   type Output = Self;
+///   fn not(self) -> Self {
+///     // ...
+///     todo!()
+///   }
+/// }
+///
+/// impl PredicateSymbol for Check {
+///   fn and(self, other: Self) -> ControlFlow<Self, Self> {
+///     // ...
+///     todo!()
+///   }
+///
+///   fn or(self, other: Self) -> ControlFlow<Self, Self> {
+///     // ...
+///     todo!()
+///   }
+/// }
+///
+/// fn shallow_eq(left: &Check, right: &Check) -> bool {
+///    match (left, right) {
+///      (Check::Good, Check::Good) | (Check::Bad(_), Check::Bad(_)) => true,
+///      _ => false
+///    }
+/// }
+///
+/// fn check_conformity() {
+///    let good = Check::Good;
+///    let bad = Check::Bad("example".to_owned());
+///    // Would fail some assertions, since PartialEq is "deep"
+///    // test_predicate_symbol_conformity!(Check; good, bad);
+///
+///    // Works as expected
+///    test_predicate_symbol_conformity!(Check; shallow_eq; good, bad);
+/// }
+/// ```
+///
+#[macro_export]
+macro_rules! test_predicate_symbol_conformity {
+    // With default eq implementation
+    ($typ:ty; $($syms:ident),*) => {
+        test_predicate_symbol_conformity!($typ; PartialEq::eq; $($syms),*);
+    };
+    // With custom shallow_eq implementation
+    ($typ:ty; $shallow_eq:path; $($syms:ident),*) => {{
+        use core::ops::Not;
+        use $crate::PredicateSymbol;
+
+        test_predicate_symbol_conformity!(__inner => $typ; $shallow_eq; $($syms),*);
+    }};
+    (__inner => $typ:ty; $eq:path; $sym:ident, $($syms:ident),+) => {
+        // Double negation elimination
+        assert!($eq(&$sym.clone().not().not(), &$sym.clone()));
+
+        // Associativity for and
+        $(assert!($eq(
+            &<$typ as PredicateSymbol>::unwrapped_and(
+                $sym.clone(),
+                $syms.clone()
+            ),
+            &<$typ as PredicateSymbol>::unwrapped_and(
+                $syms.clone(),
+                $sym.clone()
+            )
+        ));)*
+
+        // Associativity for or
+        $(assert!($eq(
+            &<$typ as PredicateSymbol>::unwrapped_or(
+                $sym.clone(),
+                $syms.clone()
+            ),
+            &<$typ as PredicateSymbol>::unwrapped_or(
+                $syms.clone(),
+                $sym.clone()
+            )
+        ));)*
```

Comment:
I'm not sure that we are testing `associativity` here, because associativity is `(a && b) && c == a && (b && c)`.
And we are testing `a && b == b && a` which is `commutativity`.

Reply:
Oh right 🤦‍♂️
I think I mixed them up basically everywhere I used the term, I'll fix it

Reply:
Also i think this testing is not exhaustive because we don't test **every** pair of elements first element is always used.

Reply:
We recur without the first symbol at the end

Reply:
We need tests for `associativity` because when we adding thing to `NonEmpty` we are relying on `(a && b) && c == a && b && c`.

Reply:
Hmm, true enough
I'll add them as well

Reply:
> We recur without the first symbol at the end

Indeed, my bad, i missed recursion.

Reply:
@Erigara fixed both issues.

<details> 
<summary> Also I realized there was absolutely no need to make this a macro.</summary>
<img src="https://i.kym-cdn.com/entries/icons/original/000/040/653/goldblum-quote.jpeg" />
</details>